R/bundlePackage.R

```diff
@@ -25,135 +25,59 @@ bundlePackages <- function(appDir,
   )
   deps <- snapshotRDependencies(appDir, inferredRDependencies, verbose = verbose)
 
-  packageMessages <- c()
-  errorMessages <- c()
-  packages <- list()
-
-  # construct package list from dependencies
-  for (i in seq_len(nrow(deps))) {
-    name <- deps[i, "Package"]
-
-    # get package info
-    info <- as.list(deps[i, c("Source", "Repository")])
-
-    # include github package info
-    info <- c(info, as.list(deps[i, grep("Github", colnames(deps), perl = TRUE, value = TRUE)]))
-
-    # get package description; note that we need to remove the
-    # packageDescription S3 class from the object or jsonlite will refuse to
-    # serialize it when building the manifest JSON
-    # TODO: should we get description from packrat/desc folder?
-    info$description <- suppressWarnings(unclass(utils::packageDescription(name)))
-
-    # TODO(HW): should this code be removed? I don't see how it is ever reached
-    # if description is NA, application dependency may not be installed
-    if (is.na(info$description[1])) {
-      errorMessages <- c(
-        errorMessages,
-        paste0(
-          capitalize(assetTypeName), " depends on package \"", name, "\"",
-          "but it is not installed. Please resolve before continuing."
-        )
-      )
-      next
-    }
-
-    # validate package source (returns a message if there is a problem)
-    packageMessages <- c(packageMessages, validatePackageSource(deps[i, ]))
-
-    # good to go
-    packages[[name]] <- info
+  not_installed <- !vapply(deps$Package, is_installed, logical(1))
+  if (any(not_installed)) {
+    pkgs <- deps$Package[not_installed]
+    cli::cli_abort(c(
+      "All packages used by {assetTypeName} must be installed.",
+      x = "Missing packages: {.pkg {pkgs}}."
+    ))
   }
 
-  if (length(packageMessages)) {
-    # Advice to help resolve installed packages that are not available using the
-    # current set of configured repositories. Each package with a missing
-    # repository has already been printed (see snapshotDependencizes).
-    #
-    # This situation used to trigger an error (halting deployment), but was
-    # softened because:
-    #   * CRAN-archived packages are not visible to our available.packages
-    #     scanning.
-    #   * Source-installed packages may be available after a manual server-side
-    #     installation.
-    #
-    # That said, an incorrectly configured "repos" option is almost always the
-    # cause.
-    packageMessages <- c(
-      packageMessages,
-      paste0(
-        "Unable to determine the source location for some packages. ",
-        "Packages should be installed from a package repository like ",
-        "CRAN or a version control system. Check that ",
-        "options('repos') refers to a package repository containing ",
-        "the needed package versions."
+  missing_repo <- !isSCMSource(deps$Source) & is.na(deps$Repository)
```

Comment:
Is it worth warning about `missing_repo` packages before aborting due to `not_installed` packages? And is there a reason this is a `warn` rather than an `abort`?

Reply:
The reason for the warn, not abort (from @aronatkins), is it's not uncommon for CRAN packages to be installed in such a way that this information is lost (e.g. installed locally or from source). But maybe I can tackle this in pursuit of #659 — I could check the package name against a list of CRAN packages and error if it's not found.

Reply:
@toph-allen The additional context is from #511 / #508, which relaxed the decisions in #412

https://github.com/rstudio/rsconnect/blob/3bf50f095e80fc58ea3b966cd40d925e111fdbbc/R/bundlePackage.R#L69-L81